subprojects/docs/src/docs/userguide/groovy_build_script_primer.adoc

```diff
@@ -0,0 +1,224 @@
+// Copyright 2018 the original author or authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[groovy_build_script_primer]]
+= A Groovy Build Script Primer
```

Comment:
It would be nice to cover how to interact with containers.
`tasks`, `configurations`, `sourceSets` etc.. and the ones contributed by plugins are based on the same abstraction and the Groovy DSL provides sugar to interact with them.

Reply:
What sorts of things are you thinking about beyond the properties and methods documented in the API docs? Happy to add anything that helps interpret what's happening in a build script.

Reply:
Right, there are reasonable API docs for the containers. I think that given these containers of named domain objects are very frequent in the DSL and that community plugins also make use of these containers to build their DSL we could consider explaining their usage in a primer.

Reply:
Do you have specific examples of what to cover? I recognise that it's important to cover domain object containers somewhere, but is there anything specific to the Groovy DSL that is worth tackling here?

Reply:
@eskatos Wondering if you have any feedback on ☝️

Reply:
@pledbrook, sorry for not following up earlier

The Groovy DSL adds some magic on top of the container API.

```groovy
def container = container(Something)
container.elementName // access element by name
container.configure {
    foo { // maybeCreate element by name and configure it
        gizmo = "else"
    }
}
```

and for polymorphic containers too.

The Kotlin DSL provides different conveniences as shown for the `tasks` container in the [migration guide](https://guides.gradle.org/migrating-build-logic-from-groovy-to-kotlin/#configuring-tasks).

But maybe this should go into a place dedicated to interacting with containers instead of here, showing both Groovy and Kotlin samples.

Reply:
No problem. Although there is some magic here, it takes the form of normal syntax: property access and code blocks/closures. It may be worth including some general info on dynamic properties and methods, but keep the specifics with the documentation for the relevant features. Of course, we'd need to add something on domain object containers because I don't think there is anything right now.

Reply:
I've raised [an issue](https://github.com/gradle/dotorg-docs/issues/288) about this.